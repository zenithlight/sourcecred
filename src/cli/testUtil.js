// @flow

import {type Command, type ExitCode, handlingErrors} from "./command";

// Run a command, capturing its stdout, stderr, and exit code. A thrown
// exception will be handled as with `handlingErrors`.
export async function run(
  command: Command,
  args: $ReadOnlyArray<string>
): Promise<{|
  +exitCode: ExitCode,
  +stdout: $ReadOnlyArray<string>,
  +stderr: $ReadOnlyArray<string>,
|}> {
  const stdout = [];
  const stderr = [];
  const exitCode = await handlingErrors(command)(args, {
    out: (x) => void stdout.push(x),
    err: (x) => void stderr.push(x),
  });
  return {exitCode, stdout, stderr};
}
