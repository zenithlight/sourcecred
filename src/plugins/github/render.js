// @flow

import React, {type Node as ReactNode} from "react";
import * as R from "./relationalView";

import Link from "../../webutil/Link";

function EntityUrl(props) {
  return (
    <Link href={props.entity.url()} target="_blank" rel="nofollow noopener">
      {props.children}
    </Link>
  );
}

function repo(x: R.Repo) {
  return (
    <EntityUrl entity={x}>
      {x.owner()}/{x.name()}
    </EntityUrl>
  );
}

function hyperlinkedNumber(x: R.Issue | R.Pull) {
  return <EntityUrl entity={x}>#{x.number()}</EntityUrl>;
}

function issue(x: R.Issue) {
  return (
    <span>
      {hyperlinkedNumber(x)}: {x.title()}
    </span>
  );
}

function pull(x: R.Pull) {
  const additions = <span style={{color: "green"}}>+{x.additions()}</span>;
  const deletions = <span style={{color: "red"}}>âˆ’{x.deletions()}</span>;
  const diff = (
    <span>
      ({additions}/{deletions})
    </span>
  );
  return (
    <span>
      {hyperlinkedNumber(x)} {diff}: {x.title()}
    </span>
  );
}

function review(x: R.Review) {
  const leader = <EntityUrl entity={x}>review</EntityUrl>;
  return (
    <span>
      {leader} on {description(x.parent())}
    </span>
  );
}

function comment(x: R.Comment) {
  const leader = <EntityUrl entity={x}>comment</EntityUrl>;
  return (
    <span>
      {leader} on {description(x.parent())}
    </span>
  );
}

function userlike(x: R.Userlike) {
  return <EntityUrl entity={x}>@{x.login()}</EntityUrl>;
}

function commit(x: R.Commit) {
  const shortHash = x.hash().slice(0, 7);
  return (
    <span>
      <EntityUrl entity={x}>{shortHash}</EntityUrl>
    </span>
  );
}

export function description(e: R.Entity): ReactNode {
  const handlers = {
    repo,
    issue,
    pull,
    review,
    comment,
    commit,
    userlike,
  };
  return R.match(handlers, e);
}
