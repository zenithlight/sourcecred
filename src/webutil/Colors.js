// @flow

import deepFreeze from "deep-freeze";
export type HexColor = string;

export default (deepFreeze({
  brand: {
    medium: "#0872A2",
    dark: "#3A066A",
  },
  accent: {
    medium: "#FF3201",
  },
}): {|
  +brand: {|
    +medium: HexColor,
    +dark: HexColor,
  |},
  +accent: {|
    +medium: HexColor,
  |},
|});
