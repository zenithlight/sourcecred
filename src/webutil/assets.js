// @flow

import normalize from "../util/pathNormalize";

/**
 * Resolver for static assets (e.g., images, PDFs) and API data (e.g.,
 * the repository registry, plugin data). Any references to resources
 * should be resolved through this API.
 */
export class Assets {
  +_root: ?string;

  /**
   * Construct a resolver given a path to the root of the site. This can
   * be a relative path, like `../..`, or an absolute path, like `/`.
   */
  constructor(root: ?string) {
    this._root = root == null ? root : normalize(root);
  }

  _getRoot(): string {
    if (this._root == null) {
      throw new Error("asset root path uninitialized");
    }
    return this._root;
  }

  /**
   * Resolve a path as if the current directory and "/" both represent
   * the site root. For instance, "foo", "/foo", and "./foo" all
   * represent the same file. It is an error to specify a file that is
   * above the root, like "../bad".
   */
  resolve(path: string): any {
    if (normalize(path.replace(/^\/+/, "")).startsWith("..")) {
      // It doesn't make sense to traverse past the site's root. This is
      // likely an error in the caller.
      throw new Error("path outside site root: " + path);
    }
    return normalize(`${this._getRoot()}/${path}`);
  }
}

/**
 * Given an absolute path `p`, return a relative path `r` such that a
 * web page at pathname `p` should use `r` to refer to the root of the
 * application. The result will only contain components `.` and `..`.
 *
 * Examples:
 *
 *   - "/foo/" maps to "..";
 *   - "/foo/bar" also maps to "..";
 *   - "/foo/bar/" maps to "../..";
 *   - "/" maps to ".".
 *
 * If the argument does not start with "/", an error will be thrown.
 */
export function rootFromPath(path: string): any {
  const normalized = normalize(path);
  if (normalized[0] !== "/") {
    throw new Error("expected absolute path: " + JSON.stringify(path));
  }
  const levels = (normalized.match(/\//g) || []).length;
  return normalize(new Array(levels - 1).fill("..").join("/"));
}
