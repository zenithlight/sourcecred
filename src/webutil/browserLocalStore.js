// @flow

import {LocalStore} from "./localStore";

/*
 * A simple abstraction over 'localStorage' to provide transparent JSON
 * serialization and deserialization.
 *
 * The implementation is borrowed heavily from Khan Academy's LocalStore
 * module, and also KaVideoPlayer's SafeLocalStore module.
 */
export default class BrowserLocalStore implements LocalStore {
  version: string;
  keyPrefix: string;

  constructor({version, keyPrefix}: {|+version: string, +keyPrefix: string|}) {
    this.version = version;
    this.keyPrefix = keyPrefix;
  }

  cacheKey(key: string): string {
    if (!key) {
      throw new Error("Falsy key provided to cacheKey: " + key);
    }
    return [this.keyPrefix, this.version, key].join(":");
  }

  get<T>(key: string, whenUnavailable: T): T {
    if (!this.isEnabled()) {
      return whenUnavailable;
    }
    try {
      const data = window.localStorage[this.cacheKey(key)];
      if (data) {
        return JSON.parse(data);
      } else {
        return whenUnavailable;
      }
    } catch (e) {
      // If we had trouble retrieving, like FF's NS_FILE_CORRUPTED:
      // http://stackoverflow.com/q/18877643/
      return whenUnavailable;
    }
  }

  set(key: string, data: mixed): void {
    if (!this.isEnabled()) {
      return;
    }
    const stringified = JSON.stringify(data);

    window.localStorage[this.cacheKey(key)] = stringified;
  }

  /*
   * Delete whatever data was associated with the given key.
   */
  del(key: string): void {
    if (!this.isEnabled()) {
      return;
    }
    const cacheKey = this.cacheKey(key);
    if (cacheKey in window.localStorage) {
      // (IE throws when deleting a non-existent entry.)
      delete window.localStorage[cacheKey];
    }
  }

  /*
   * Local storage might be disabled in old browsers or in Safari's
   * private browsing mode. Don't die.
   */
  isEnabled(): boolean {
    const uid = String(+new Date());
    try {
      window.sessionStorage[uid] = uid;
      const enabled = window.sessionStorage[uid] === uid;
      window.sessionStorage.removeItem(uid);
      return enabled;
    } catch (e) {
      return false;
    }
  }
}
