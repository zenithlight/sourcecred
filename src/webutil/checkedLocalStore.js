// @flow

import deepEqual from "lodash.isequal";

import {LocalStore} from "./localStore";

/*
 * A wrapper for a `LocalStore` that adds contract checking. This
 * implementation verifies that all provided keys are strings and all
 * provided values are plain old JSON values.
 */
export default class CheckedLocalStore implements LocalStore {
  _base: LocalStore;

  constructor(base: LocalStore): void {
    this._base = base;
  }

  get<T>(key: string, whenUnavailable: T): T {
    this._validateKey(key);
    return this._base.get(key, whenUnavailable);
  }

  set(key: string, data: mixed): void {
    this._validateKey(key);
    this._validateValue(data);
    return this._base.set(key, data);
  }

  del(key: string): void {
    this._validateKey(key);
    return this._base.del(key);
  }

  _validateKey(key: string) {
    if (typeof key !== "string") {
      throw new Error(`bad key (${typeof key}): ${key}`);
    }
  }

  _validateValue(data: any) {
    try {
      if (deepEqual(data, JSON.parse(JSON.stringify(data)))) {
        return;
      }
    } catch (_) {} // eslint-disable-line no-empty
    throw new Error(`bad value: ${data}`);
  }
}
