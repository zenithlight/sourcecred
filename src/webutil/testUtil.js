// @flow

export function configureEnzyme() {
  const Enzyme = require("enzyme");
  const Adapter = require("enzyme-adapter-react-16");
  Enzyme.configure({adapter: new Adapter()});
  beforeEach(() => {
    // $FlowExpectedError[cannot-write]
    console.error = jest.fn();
    // $FlowExpectedError[cannot-write]
    console.warn = jest.fn();
  });
  afterEach(() => {
    expect(console.warn).not.toHaveBeenCalled();
    expect(console.error).not.toHaveBeenCalled();
  });
}

export function configureAphrodite() {
  const {StyleSheetTestUtils} = require("aphrodite/no-important");
  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });
}
