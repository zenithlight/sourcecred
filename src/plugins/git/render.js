// @flow

import React, {type Element as ReactElement} from "react";
import Link from "../../webutil/Link";
import * as N from "./nodes";
import type {Repository} from "./types";
import {type RepoIdString, stringToRepoId} from "../github/repoId";
import type {GitGateway} from "./gitGateway";

export function description(
  address: N.StructuredAddress,
  repository: Repository,
  gateway: GitGateway
): ReactElement<"code"> | ReactElement<"span"> {
  switch (address.type) {
    case "COMMIT": {
      const hash = address.hash;
      const commit = repository.commits[hash];
      if (commit == null) {
        return <code>{hash}</code>;
      }
      // This `any`-cast courtesy of facebook/flow#6927.
      const repoIdStrings: $ReadOnlyArray<RepoIdString> = (Object.keys(
        repository.commitToRepoId[hash] || {}
      ): any);
      if (repoIdStrings.length === 0) {
        console.error(`Unable to find repoIds for commit ${hash}`);
        // TODO(@wchargin): This shortHash is unambiguous for a single repo,
        // but might be ambiguous across many repositories. Consider disambiguating
        return (
          <span>
            <code>{commit.shortHash}</code>: {commit.summary}
          </span>
        );
      }
      const repoId = stringToRepoId(repoIdStrings[0]);
      const url = gateway.commitUrl(repoId, commit.hash);
      const hyperlinkedHash = hyperlink(url, commit.shortHash);
      return (
        <span>
          <code>{hyperlinkedHash}</code>: {commit.summary}
        </span>
      );
    }
    default:
      throw new Error(`unknown type: ${(address.type: empty)}`);
  }
}

function hyperlink(url, text) {
  return (
    <Link href={url} target="_blank" rel="nofollow noopener">
      {text}
    </Link>
  );
}
