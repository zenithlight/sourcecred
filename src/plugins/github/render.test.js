// @flow

import {render} from "enzyme";
import {exampleEntities} from "./example/example";
import {description} from "./render";
import enzymeToJSON from "enzyme-to-json";

require("../../webutil/testUtil").configureAphrodite();
require("../../webutil/testUtil").configureEnzyme();

describe("plugins/github/render", () => {
  const examples = exampleEntities();
  for (const name of Object.keys(examples)) {
    it(`renders the right description for a ${name}`, () => {
      const entity = examples[name];
      const renderedEntity = render(description(entity));
      expect(enzymeToJSON(renderedEntity)).toMatchSnapshot();
    });
  }
});
