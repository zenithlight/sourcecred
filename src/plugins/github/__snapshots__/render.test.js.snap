// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/github/render renders the right description for a comment 1`] = `
<span>
  <a
    class="link_1cmobhd"
    href="https://github.com/sourcecred-test/example-github/pull/5#discussion_r171460198"
    rel="nofollow noopener"
    target="_blank"
  >
    comment
  </a>
   on 
  <span>
    <a
      class="link_1cmobhd"
      href="https://github.com/sourcecred-test/example-github/pull/5#pullrequestreview-100313899"
      rel="nofollow noopener"
      target="_blank"
    >
      review
    </a>
     on 
    <span>
      <a
        class="link_1cmobhd"
        href="https://github.com/sourcecred-test/example-github/pull/5"
        rel="nofollow noopener"
        target="_blank"
      >
        #5
      </a>
       
      <span>
        (
        <span
          style="color:green"
        >
          +1
        </span>
        /
        <span
          style="color:red"
        >
          −0
        </span>
        )
      </span>
      : This pull request will be more contentious. I can feel it...
    </span>
  </span>
</span>
`;

exports[`plugins/github/render renders the right description for a commit 1`] = `
<span>
  <a
    class="link_1cmobhd"
    href="https://github.com/sourcecred-test/example-github/commit/0a223346b4e6dec0127b1e6aa892c4ee0424b66a"
    rel="nofollow noopener"
    target="_blank"
  >
    0a22334
  </a>
</span>
`;

exports[`plugins/github/render renders the right description for a issue 1`] = `
<span>
  <a
    class="link_1cmobhd"
    href="https://github.com/sourcecred-test/example-github/issues/2"
    rel="nofollow noopener"
    target="_blank"
  >
    #2
  </a>
  : A referencing issue.
</span>
`;

exports[`plugins/github/render renders the right description for a pull 1`] = `
<span>
  <a
    class="link_1cmobhd"
    href="https://github.com/sourcecred-test/example-github/pull/5"
    rel="nofollow noopener"
    target="_blank"
  >
    #5
  </a>
   
  <span>
    (
    <span
      style="color:green"
    >
      +1
    </span>
    /
    <span
      style="color:red"
    >
      −0
    </span>
    )
  </span>
  : This pull request will be more contentious. I can feel it...
</span>
`;

exports[`plugins/github/render renders the right description for a repo 1`] = `
<a
  class="link_1cmobhd"
  href="https://github.com/sourcecred-test/example-github"
  rel="nofollow noopener"
  target="_blank"
>
  sourcecred-test/example-github
</a>
`;

exports[`plugins/github/render renders the right description for a review 1`] = `
<span>
  <a
    class="link_1cmobhd"
    href="https://github.com/sourcecred-test/example-github/pull/5#pullrequestreview-100313899"
    rel="nofollow noopener"
    target="_blank"
  >
    review
  </a>
   on 
  <span>
    <a
      class="link_1cmobhd"
      href="https://github.com/sourcecred-test/example-github/pull/5"
      rel="nofollow noopener"
      target="_blank"
    >
      #5
    </a>
     
    <span>
      (
      <span
        style="color:green"
      >
        +1
      </span>
      /
      <span
        style="color:red"
      >
        −0
      </span>
      )
    </span>
    : This pull request will be more contentious. I can feel it...
  </span>
</span>
`;

exports[`plugins/github/render renders the right description for a userlike 1`] = `
<a
  class="link_1cmobhd"
  href="https://github.com/wchargin"
  rel="nofollow noopener"
  target="_blank"
>
  @wchargin
</a>
`;
